File: Haskell/Day04.hs
import AdventHelper

import Data.List
import Data.List.Split

import Text.Regex.TDFA
import Text.Regex.TDFA.Text ()

complete = ["byr","cid","ecl","eyr","hcl","hgt","iyr","pid"]

checkId1 :: String -> Bool
checkId1 s = length f == 0
  where s' = splitOn " " s
        cs = sort $ map (takeWhile (/= ':')) s'
        f = (complete \\ cs) \\ ["cid"]

checkId2 :: String -> Bool
checkId2 s = all (checkUnit) us
  where s' = splitOn " " s
        us = map (\x -> tuplify2 $ splitOn ":" x) s'

checkUnit :: (String, String) -> Bool
checkUnit ("byr",v) = v =~ "(19[2-9][0-9]|200[0-2])"
checkUnit ("iyr",v) = v =~ "(201[0-9]|2020)"
checkUnit ("eyr",v) = v =~ "(202[0-9]|2030)"
checkUnit ("hgt",v) = v =~ "(59in|6[0-9]in|7[0-6]in|1[5-8][0-9]cm|19[0-3]cm)"
checkUnit ("hcl",v) = v =~ "^#([0-9a-f]){6}$"
checkUnit ("ecl",v) = v =~ "(amb|blu|brn|gry|grn|hzl|oth)"
checkUnit ("pid",v) = v =~ "^([0-9]){9}$"
checkUnit ("cid",v) = True
checkUnit otherwise = error "Unidentified category"

main = do
  putStrLn "Day 4"
  f <- readFile "../input/input04.txt"
  let s = parseLineGroups " " $ lines f

  printSoln 1 (length $ filter (checkId1) s)
  printSoln 2 (length $ filter (checkId2) $ filter (checkId1) s)

File: C++/Day04.cpp
#include "AH.h"

namespace Day04
{

	std::tuple<bool, bool> check_id(const std::string& s)
	{
		
		bool part1 = true, part2 = true;
		auto parts = AH::Split(s, ' ');

		std::set<std::string> seen;
		std::vector<std::string> fields{"byr", "iyr", "eyr", "hgt",
		                                "hcl", "ecl", "pid"};

		for (auto pw : parts)
		{
		  	auto p = AH::Split(pw, ':');
		  	auto key = p[0];
		  	auto value = p[1];
				seen.insert(p[0]);

	  		std::smatch m;

			if (key == "byr")
				part2 &= (std::regex_search (value, m, std::regex("(19[2-9][0-9]|200[0-2])")));
			else if (key == "iyr")
				part2 &= (std::regex_search (value, m, std::regex("(201[0-9]|2020)")));
			else if (key == "eyr")
				part2 &= (std::regex_search (value, m, std::regex("(202[0-9]|2030)")));
			else if (key == "hgt")
				part2 &= (std::regex_search (value, m,
				                             std::regex("(59in|6[0-9]in|7[0-6]in|1[5-8][0-9]cm|19[0-3]cm)")));
			else if (key == "hcl")
				part2 &= (std::regex_search (value, m, std::regex("#([0-9a-f]){6}")));
			else if (key == "ecl")
				part2 &= (std::regex_search (value, m,
				                             std::regex("(amb|blu|brn|gry|grn|hzl|oth)")));
			else if (key == "pid")
				part2 &= (std::regex_search (value, m, std::regex("^([0-9]){9}$")));
		}

		for (auto f : fields)
	  		part1 &= (seen.find(f) != seen.end());

		return std::make_tuple( part1, part1 & part2);

	}

	int Run(const std::string& filename)
	{
		auto inputLines = AH::ReadTextFile(filename);
		auto lineGroups = AH::ParseLineGroups(inputLines);

		uint64_t p1 = 0, p2 = 0;
		for (auto pw : lineGroups)
		{
			auto [ok1, ok2] = check_id(pw);
			p1 += ok1 ? 1 : 0;
			p2 += ok2 ? 1 : 0;
		}

		AH::PrintSoln(4, p1, p2);

		return 0;
	}

}
